code/modules/surgery/surgery.dm

```diff
@@ -41,146 +137,351 @@
+	// 	to_chat(user, "<span class='warning'>[target] seems to be missing the organ necessary to complete this surgery!</span>")
+	// 	return FALSE
+
+	if(is_valid_tool(user, tool))
 		if(target_zone == surgery.location)
-			initiate(user, target, target_zone, tool, surgery)
+			if(get_location_accessible(target, target_zone) || surgery.ignore_clothes)
+				initiate(user, target, target_zone, tool, surgery)
+			else
+				to_chat(user, "<span class='warning>You need to expose [target]'s [parse_zone(target_zone)] before you can perform surgery on it!")
 			return TRUE //returns TRUE so we don't stab the guy in the dick or wherever.
+
+	if(repeatable)
+		// you can continuously, manually, perform a step, so long as you continue to use the correct tool.
+		// if you use the wrong tool, though, it'll try to start the next surgery step with the tool you have in your hand.
+
+		// Note that this separate from returning the surgery_step_* defines in end/fail_step, which will automatically retry the surgery.
+
+
+
+		// TODO Make sure we have support for it!
+		var/datum/surgery_step/next_step = surgery.get_surgery_next_step()
+		next_step.times_repeated = times_repeated + 1
+		if(next_step)
+			surgery.status++
+			if(next_step.try_op(user, target, user.zone_selected, user.get_active_hand(), surgery))
+				return TRUE
+			else
+				surgery.status--
+
+	return FALSE
+
+/**
+ * Determines whether or not this surgery step can repeat if its end/fail steps returned SURGERY_STEP_RETRY.
+ *
+ * Arguments:
+ * * user - mob performing the surgery
+ * * target - mob the surgery is being performed on
+ * * target_zone - body zone of the surgery
+ * * tool - tool used for the surgery
+ * * surgery - the operation this surgery step is a part of
+ *
+ * If this returns TRUE, the step will automatically retry. If not, the user will have to manually start the step again.
+ *
+ */
+/datum/surgery_step/proc/can_repeat(mob/living/user, mob/living/carbon/target, target_zone, obj/item/tool, datum/surgery/surgery)
+	if(tool && istype(tool) && (tool.flags & SURGICALTOOL_ADVANCED))
+		return TRUE
+	if(HAS_TRAIT(user, TRAIT_REPEATSURGERY))
+		return TRUE
 	return FALSE
 
-/datum/surgery_step/proc/initiate(mob/user, mob/living/carbon/target, target_zone, obj/item/tool, datum/surgery/surgery)
+/**
+ * Initiate and really perform the surgery itself.
+ * This includes the main do-after and the checking of probabilities for successful surgeries.
+ * If try_to_fail is TRUE, then this surgery will be deliberately failed out of.
+ *
+ * Returns TRUE if the surgery should proceed to the next step, or FALSE otherwise.
+ */
+/datum/surgery_step/proc/initiate(mob/user, mob/living/carbon/target, target_zone, obj/item/tool, datum/surgery/surgery, try_to_fail = FALSE)
+
+	// TODO This proc isn't really used by /tg/, and might cause some weird conflicts between surgeries and surgery steps if we aren't careful.
 	if(!can_use(user, target, target_zone, tool, surgery))
 		return
```

Comment:
I really can't decide if this proc should stay or go. It's historically been used to add some limitations to how surgery steps can be done, but really doesn't make much sense in its implementation. If you're running through a linear surgery, and you get to a step that has `can_use == FALSE`, what do you do? Realistically, the checks that happen in this step should go into the `begin_step` or `tool_check`. (For the record, this proc isn't present in /tg/'s implementation).

That being *said*, I can maybe see there being merit in leaving these in place for non-linear surgeries, so you don't have to duplicate checks, but it still feels messy. I think it can go, just not entirely sure yet.

Reply:
Never mind. Decided we were better off without it.